[DAGCombiner] Add combine for vector interleave of splats

david-arm opened the pull request:
This patch adds two DAG combines:

1. vector_interleave(splat, splat, ...) -> {splat,splat,...}
2. concat_vectors(splat, splat, ...) -> wide_splat

where all the input splats are identical. Both of these
together enable us to fold
  concat_vectors(vector_interleave(splat, splat, ...))
into a wide splat. Post-legalisation we must only do the
concat_vector combine if the wider type and splat operation
is legal.

For fixed-width vectors the DAG combine only occurs for
interleave factors of 3 or more, however it's not currently
safe to test this for AArch64 since there isn't any lowering
support for fixed-width interleaves. I've only added
fixed-width tests for RISCV.

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -12594,27 +12594,40 @@ void SelectionDAGBuilder::visitVectorDeinterleave(const CallInst &I,
+      OperandsAreSame = false;
+  }
+
+  if (OperandsAreSame) {
+    if (auto *C = dyn_cast<Constant>(CI.getOperand(0))) {
+      if (auto *SV = C->getSplatValue()) {
+        SDValue Res = DAG.getNode(ISD::SPLAT_VECTOR, DL, OutVT, getValue(SV));
+        setValue(&CI, Res);
+        return;
+      }
+    }
```
I was thinking more if there were any more SPLAT_VECTORs formed it might be useful to also catch those. But I'm not sure how common that is

paulwalker-arm replied:
Current applicability doesn't really matter.  As @lukel97 says, there's the possibility that the splat might not be present during initial construction but appear later after a series of combines and/or legalisation.  For this reason it is better to implement this as a  DAG combine to maximise any future utility.

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25173,6 +25173,38 @@ static SDValue combineConcatVectorOfCasts(SDNode *N, SelectionDAG &DAG) {
+      return SDValue();
+  }
+
+  SDValue InOp0 = FirstOp.getOperand(0);
+  if (!llvm::all_of(FirstOp->ops(),
+                    [&InOp0](SDValue Op) { return Op == InOp0; }))
+    return SDValue();
```
Good suggestion, I'll give it a go!

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25397,6 +25429,9 @@ SDValue DAGCombiner::visitCONCAT_VECTORS(SDNode *N) {
     return DAG.getBuildVector(VT, SDLoc(N), Opnds);
   }
 
+  if (SDValue V = combineConcatVectorInterleave(N, DAG))
+    return V;
+
```
In that case is it really a DAG combine? It sounds more like a simplification of a node, which might be better done at the time of creation in SelectionDAG::getNode perhaps? It could work, but I'm not sure how to do it because the node still needs to set N results. I see what you mean though - if only the first result gets used then this DAG combine doesn't help. I kind of expected the common use case at the moment is the pattern created by SelectionDAGBuilder when lowering the llvm.vector.interleave intrinsic. @paulwalker-arm any thoughts?

paulwalker-arm replied:
Relying on the concat will unnecessarily restrict the combine making it unlikely to fire after type legalisation so I'd rather this be an interleave specific combine without caring about its uses.  Which I think will simplify the code as well?

Doing this in `SelectionDAG::getNode()` sounds reasonable but then the combiner doesn't always create new nodes so having the combine could catch more case. These are multi-result nodes so perhaps just choose the path with the cleanest implementation (I'll guess DAGCombiner will win here).

david-arm replied:
OK, I've tried the suggestion of doing the combine on the VECTOR_INTERLEAVE only, but it also required adding a separate combine for concat_vector(splat, splat, ...).

david-arm commented on llvm/test/CodeGen/RISCV/rvv/vector-interleave.ll:
```diff
@@ -14947,3 +14947,183 @@ define <vscale x 16 x double> @vector_interleave_nxv16f64_nxv2f64(<vscale x 2 x
+}
+
+define <vscale x 4 x i16> @interleave2_diff_nonconst_splat_nxv4i16(i16 %a, i16 %b) {
+; SVE-LABEL: interleave2_diff_nonconst_splat_nxv4i16:
+; SVE:       // %bb.0:
+; SVE-NEXT:    // kill: def $w1 killed $w1 def $x1
+; SVE-NEXT:    // kill: def $w0 killed $w0 def $x0
+; SVE-NEXT:    mov z0.d, x0
+; SVE-NEXT:    mov z1.d, x1
+; SVE-NEXT:    zip2 z2.d, z0.d, z1.d
+; SVE-NEXT:    zip1 z0.d, z0.d, z1.d
+; SVE-NEXT:    uzp1 z0.s, z0.s, z2.s
+; SVE-NEXT:    ret
+;
+; SME2-LABEL: interleave2_diff_nonconst_splat_nxv4i16:
```
Ha, the result of lazy copy and paste. :)

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25173,6 +25173,37 @@ static SDValue combineConcatVectorOfCasts(SDNode *N, SelectionDAG &DAG) {
+      FirstOp.getNumOperands() != N->getNumOperands())
+    return SDValue();
+
+  for (unsigned I = 0; I < N->getNumOperands(); I++) {
```
Nit, does enumerating op_values also work here?
```suggestion
  for (auto [Idx, Op] : enumerate(N->op_values())) {
```

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25173,6 +25173,37 @@ static SDValue combineConcatVectorOfCasts(SDNode *N, SelectionDAG &DAG) {
+  SDLoc DL(N);
+  EVT SubVecTy = InOp0.getValueType();
+  // Create the wider type required.
+  EVT WideVecTy =
+      EVT::getVectorVT(*DAG.getContext(), SubVecTy.getScalarType(),
+                       SubVecTy.getVectorElementCount() * N->getNumOperands());
```
Is WideVecTy is the same as the concat_vector's type? Can we just replace this with N->getValueType(0) instead?

lukel97 reviewed (approved):
LGTM

mshockwave commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25173,6 +25173,37 @@ static SDValue combineConcatVectorOfCasts(SDNode *N, SelectionDAG &DAG) {
   return DAG.getNode(CastOpcode, DL, VT, NewConcat);
 }
 
+static SDValue combineConcatVectorInterleave(SDNode *N, SelectionDAG &DAG) {
```
could you add a comment here to briefly explain your pattern?

lukel97 reviewed (approved):
New version also LGTM

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25465,6 +25502,24 @@ SDValue DAGCombiner::visitCONCAT_VECTORS(SDNode *N) {
+    return SDValue();
+
+  // Check to see if the identical operand is a splat.
+  SDValue Op0 = N->getOperand(0);
+  SDValue Splat = DAG.getSplatValue(Op0);
```
```suggestion
  SDValue Splat = DAG.getSplatValue(N->getOperand(0));
```

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25465,6 +25502,24 @@ SDValue DAGCombiner::visitCONCAT_VECTORS(SDNode *N) {
+  // Simply replace all results with the first operand.
+  SmallVector<SDValue, 4> Ops;
+  Ops.append(N->op_values().begin(), N->op_values().end());
+  SDValue Res = CombineTo(N, &Ops);
+  return Res;
```
```suggestion
  return CombineTo(N, &Ops);
```

david-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -331,6 +331,11 @@ namespace {
       return CombineTo(N, To, 2, AddTo);
     }
 
+    SDValue CombineTo(SDNode *N, SmallVectorImpl<SDValue> *To,
```
This fails to build with error:

```
DAGCombiner.cpp:25519:27: error: no matching function for call to ‘{anonymous}::DAGCombiner::CombineTo(llvm::SDNode*&, llvm::SmallVector<llvm::SDValue, 4>*)’
25519 |   return CombineTo(N, &Ops);
```

The existing interface for CombineTo isn't great:

```
    SDValue CombineTo(SDNode *N, const SDValue *To, unsigned NumTo,
                      bool AddTo = true);
```

because there is no easy way to get access to a C-like array of SDValue operands from a SDNode, which is why I went down the `SmallVector` route although it also feels a bit awkward.

david-arm committed:
Add tests

david-arm committed:
[DAGCombiner] Add combine for vector interleave of splats

This patch adds two DAG combines:

1. vector_interleave(splat, splat, ...) -> {splat,splat,...}
2. concat_vectors(splat, splat, ...) -> wide_splat

where all the input splats are identical. Both of these
together enable us to fold
  concat_vectors(vector_interleave(splat, splat, ...))
into a wide splat. Post-legalisation we must only do the
concat_vector combine if the wider type and splat operation
is legal.

For fixed-width vectors the DAG combine only occurs for
interleave factors of 3 or more, however it's not currently
safe to test this for AArch64 since there isn't any lowering
support for fixed-width interleaves. I've only added
fixed-width tests for RISCV.

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25465,6 +25498,22 @@ SDValue DAGCombiner::visitCONCAT_VECTORS(SDNode *N) {
+    return SDValue();
+
+  // Check to see if the identical operand is a splat.
+  SDValue Splat = DAG.getSplatValue(N->getOperand(0));
+  if (!Splat)
```
Sorry, I should have noticed this before but does `if (!isSplatValue(N->getOperand(0))` work here?

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -25465,6 +25498,22 @@ SDValue DAGCombiner::visitCONCAT_VECTORS(SDNode *N) {
+  if (!Splat)
+    return SDValue();
+
+  // Simply replace all results with the first operand.
```
Perhaps "interleave splat(X), splat(X).... --> splat(X), splat(X)...."?

paulwalker-arm reviewed (approved):
A final suggestion but otherwise this looks good to me.

david-arm committed:
Address review comments

david-arm commented:
I rebased and ran make check-all downstream, which seems fine.
